vpr/src/place/cut_spreader.h

```diff
@@ -0,0 +1,259 @@
+
+    // Utilization of each tile, indexed by x, y
+    std::vector<std::vector<int>> occupancy;
+
+    // Region ID of each tile, indexed by x, y. -1 if not covered by any region
+    std::vector<std::vector<int>> groups;
+
+    // Extent of macro at x, y location. If blk is not in any macros, MacroExtent only covers a single locations
+    std::vector<std::vector<MacroExtent>> macros;
+
+    // List of logic blocks of blk_type at x, y location, indexed by x, y
+    std::vector<std::vector<std::vector<ClusterBlockId>>> blks_at_location;
+
+    // List of sub_tiles compatible with blk_type at x, y location, indexed by x, y
+    std::vector<std::vector<std::vector<t_pl_loc>>> ft;
+
+    // List of all SpreaderRegion, index of vector members is the id of the region
+    std::vector<SpreaderRegion> regions;
+
+    // List of all merged_regions, these regions are merged in larger regions and should be skipped when
+    // recursively cur_spreading. Each entry is the region's ID, which is also the index into the regions vector
+    std::unordered_set<int> merged_regions;
+
+    // Lookup of macro's extent by block ID. If block is a single block, MacroExtent contains only 1 tile location
+    std::map<ClusterBlockId, MacroExtent> blk_extents;
+
+    // Setup CutSpreader data structures using information from AnalyticPlacer
+    // including blks_at_location, macros, groups, etc.
+    void init();
+
+    // Returns number of logical blocks at x, y location
+    int occ_at(int x, int y);
+
+    // Returns number of compatible sub_tiles at x, y location
+    int tiles_at(int x, int y);
+
+    /*
+     * Merge mergee into merged by:
+     * * change group id at mergee grids to merged id
+     * * adds all n_blks and n_tiles from mergee to merged region
+     * * grow merged to include all mergee grids
+     */
+    void merge_regions(SpreaderRegion& merged, SpreaderRegion& mergee);
+
+    /*
+     * Grow region r to include a rectangular region (x0, y0, x1, y1)
+     * Pass init = true if first time calling for a newly created region
+     */
+    void grow_region(SpreaderRegion& r, int x0, int y0, int x1, int y1, bool init = false);
+
+    /*
+     * Expand all utilized regions until they satisfy n_tiles * beta >= n_blocks
+     * If overutilized regions overlap in this process, they are merged
+     */
+    void expand_regions();
+
+    /*
+     * Find overutilized regions surrounded by non-overutilized regions
+     * Start off at an overutilized tile and expand in x, y directions until the region is surrounded by
+     * non-overutilized regions.
+     */
+    void find_overused_regions();
+
+    /*
+     * Recursive cut-based spreading in HeAP paper
+     * "left" denotes "-x, -y", "right" denotes "+x, +y" depending on dir
+     *
+     *  @param r	regrion to cut & spread
+     *  @param dir	direction, true for y, false for x
+     *
+     *  @return		a pair of sub-region IDs created from cutting region r.
+     *  			{-2, -2} if base case is reached
```

Comment:
Make these values (-2,-2 and -1,-1) defined constants (pairs) so you can check for them with clean looking code. Or return a separate enumerated constant status code (e.g. with a by-reference outcome variable).